ouroboros-consensus/src/Ouroboros/Storage/ChainDB/Impl/Background.hs

```diff
@@ -322,3 +333,84 @@ updateLedgerSnapshotsRunner cdb@CDB{..} = loop
     waitInterval = do
       interval <- atomically onDiskWriteInterval
       threadDelay interval
+
+{-------------------------------------------------------------------------------
+  Executing scheduled chain selections
+-------------------------------------------------------------------------------}
+
+-- | Retrieve the blocks from 'cdbFutureBlocks' for which chain selection was
+-- scheduled at the current slot /or before/. Run chain selection for each of
+-- them.
+scheduledChainSelection
+  :: (IOLike m, ProtocolLedgerView blk, HasCallStack)
+  => ChainDbEnv m blk
+  -> SlotNo  -- ^ The current slot
+  -> m ()
+scheduledChainSelection cdb@CDB{..} curSlot = do
+    -- The list of hashes is stored in reverse order so we can easily prepend.
+    -- We reverse it here so that we execute the chain selections in
+    -- chronological order, so it is likelier that we can simply append a
+    -- block to the end of the chain.
+    points <- atomically $ updateTVar cdbFutureBlocks $ \futureBlocks ->
+      let (before, at, after) = Map.splitLookup curSlot futureBlocks
+          points =
+            [ BlockPoint slot hash
+            | (slot, hashes) <- Map.toAscList before
+            , hash <- reverse (NE.toList hashes)
+            ] <>
+            [ BlockPoint curSlot hash
+            | hash <- reverse (maybe [] NE.toList at)
+            ]
+      in (after, points)
+
+    unless (null points) $ do
+      traceWith cdbTracer $
+        TraceAddBlockEvent $ RunningScheduledChainSelection points curSlot
+      forM_ points $ \point ->
+        -- If an exception occurs here, then no chain selection will be
+        -- performed for the blocks after it. Since only real errors that
+        -- would shut down the ChainDB could occur here, this is fine. When
+        -- (re)starting the ChainDB, full chain selection is performed, which
+        -- would include these blocks.
+        chainSelectionForBlock cdb (Left point)
+
+{-
+
+TODO switch back to:
+
+-- | Retrieve the blocks from 'cdbFutureBlocks' for which chain selection was
+-- scheduled at the current slot. Run chain selection for each of them.
+scheduledChainSelection
+  :: (IOLike m, ProtocolLedgerView blk, HasCallStack)
+  => ChainDbEnv m blk
+  -> SlotNo  -- ^ The current slot
+  -> m ()
+scheduledChainSelection cdb@CDB{..} curSlot = do
+    mbHashes <- atomically $ updateTVar cdbFutureBlocks $
+      -- Extract and delete the value stored at @curSlot@
+      swap . Map.updateLookupWithKey (\_ _ -> Nothing) curSlot
+    whenJust mbHashes $ \hashes -> do
+    -- The list of hashes is stored in reverse order so we can easily prepend.
```

Comment:
They're all in the same slot number, so they must all conflict; we can only adopt one.
Even for EBBs this doesn't matter, explain why.